(PE-16279) clj-ldap does not validate certname matches hostname

dmcpl opened the pull request:
Getting this PR visible to verify that the changes suggested are valid. Code is still in rough shape and will be tidied up once I find out that the changes aren't crazy.

* Added a new option :verify-host? which if true will add a HostNameSSLSocketVerifier to the LDAPConnectionOptions as suggested by @underscorgan 

* For testing I have a new namespace in which an ApacheDS server is bought up with a new self-signed keystore (stored in dev-resources/keystore-local.jks)

* The keystore was created thusly:
```
keytool -genkey -keyalg RSA -alias selfsigned -keystore keystore-local.jks -storepass password -validity 360 -keysize 2048
What is your first and last name?
  [Unknown]:  localhost
...
Is CN=localhost, OU=nah, O=nah, L=nah, ST=nah, C=US correct?
...
```

* There are some basic tests connecting with 'localhost' which passes and '127.0.0.1' which fails when the new `:verify-host` is true. Both pass when the param is false.

Is are these tests off the mark? I am still learning as I go along so any advice is welcome.

underscorgan commented:
@dmcpl I can't reasonably review this, I have limited/no clojure experience.

dmcpl commented:
@underscorgan no worries, I will get the clojure heads on it tomorrow! thanks anyways

dmcpl commented:
@underscorgan one thing you may be able to comment on. I have loaded a test ApacheDS server with a keystore containing CN=localhost and to prove hostname verification I connect with '127.0.0.1' and it blows up as expected. Is this a cheat that should be avoided? I was aware that these tests will be run on a CI server and I could provide no specific hostnames.

dmcpl commented on src/clj_ldap/client.clj:
```diff
@@ -99,10 +100,11 @@
   (let [opt (LDAPConnectionOptions.)]
     (when connect-timeout (.setConnectTimeoutMillis opt connect-timeout))
     (when timeout         (.setResponseTimeoutMillis opt timeout))
+    (when verify-host?    (.setSSLSocketVerifier opt (HostNameSSLSocketVerifier. false)))
```
The `false` flag on `HostNameSSLSocketVerifier` is for `allowWidlcards`. I wasn't sure whether to permit this. Any comments?

eoinmcq replied:
Personally I would say we should use true the allowWildCards as if a customer has set a wildcard in their cert then it could be legitimate but could fail.

An alternative would be to make this parameter configurable but seems like a lot of work for not that much gain.

jonathannewman replied:
My instinct is to make this an option that can be specified, and default to false. @jsane and @joshcooper have any thoughts around this?

jsane replied:
While there seems to be widespread industry practice of using wildcard certificates for a variety of reasons it is not endorsed or mentioned in any pkix specs. On the contrary rfc6125 discourages use of wildcards. The problem with wildcards is the matching rules are not standard and varied. It's possible that some of our customers might actually be using them and disallowing wildcards w/o any configuration option is likely to break things. 
But allowing them seems an insecure choice. I imagine (& know) it would be much easier for an insider to obtain a certificate within an organization that will match. That defeats the purpose of what this bug is trying to address. 
So an additional option to allow/disallow wildcards (default to false) would be preferable so whoever can make a conscious choice. 

If we decide to always allow wildcards then we may need to document the behavior of our wildcard matching (which could be a time consuming exercise) and provide a security warning of potential implications with it.

dmcpl commented:
Its 360 days from time of creation. I will create another cert with 20 years expiry which should (hopefully) push this problem beyond the life of this project.

dmcpl commented:
Added `:wildcard-host?` in the option map of `connect` which will allow wildcards in certs. This is off by default, and will have no effect if `:verify-host` is not also `true`.

Had to manually test these changes messing with /etc/hosts as I could not simulate wildcarded hostnames in the unit tests.

Updated docs and commit message to reflect these changes.

Some other small pieces of tidy up.
